src/main/java/com/iota/iri/service/milestone/impl/LatestSolidMilestoneTrackerImpl.java

```diff
@@ -191,4 +243,32 @@ private void logChange(int prevSolidMilestoneIndex) {
             messageQ.publish("lmhs %s", latestMilestoneHash);
         }
     }
+
+    /**
+     * Tries to actively repair the ledger by reverting the milestones preceding the given milestone.<br />
+     * <br />
+     * It gets called when a milestone could not be applied to the ledger state because of problems like "inconsistent
+     * balances". While this should theoretically never happen (because milestones are by definition "consistent"), it
+     * can still happen because IRI crashed or got stopped in the middle of applying a milestone or if a milestone
+     * was processed in the wrong order.<br />
+     * <br />
+     * Every time we call this method the internal {@link #repairBackoffCounter} gets increased which causes the next
+     * call of this method to repair an additional milestone. This means that whenever we face an error we first try to
+     * reset only the last milestone, then the two last milestones, then the three last milestones (and so on ...) until
+     * the problem was fixed.<br />
+     * <br />
+     * To be able to tell when the problem is fixed and the {@link #repairBackoffCounter} can be reset, we store the
+     * milestone index that caused the problem the first time we call this method.<br />
+     *
+     * @param errorCausingMilestone the milestone that failed to be applied
+     * @throws MilestoneException if we failed to reset the corrupted milestone
+     */
+    private void revertPrecedingMilestones(MilestoneViewModel errorCausingMilestone) throws MilestoneException {
```

Comment:
"Preceding" can be confusing, is it Preceding in the process or in index #?

Reply:
maybe `outOfOrder`?

Reply:
Both, since milestones get processed in the order that they are issued.

outOfOrder is wrong - we reset the milestones preceding the errorCausingMilestone - during this process we might identify which milestone was processed out of order (or marked the wrong transactions) but the milestones we reset here are only the ones that were issued before the once that failed to get applied.

Reply:
changed it and introduces some additional utility methods for better readability